Keep a radius of cached regions loaded

UPSman123 opened the pull request:
This implements an enhancement from #324. I created the ChunkLoader
class. This handles loading chunks around the player and render view
entity in a given radius. The chunks are loaded from disk using the
CachedWorld.tryLoadFromDisk function. If a chunk has not yet been
added to the cache (in memory or on disk) then this will not cache it.

Curently the class is initialized from the Baritone constructor. I don't know if this is okay or if it should be done elsewhere.
I couldn't find a good place for it because it needs to be registered with the GameEventHandler.

I also noticed that the ICachedWorld interface does not include the tryLoadFromDisk function even though it is public.
I don't know if this is intentional or not. In this commit I decided to just cast the ICachedWorld to a CachedWorld.

<!-- No UwU's or OwO's allowed -->

ZacSharp commented on src/main/java/baritone/cache/ChunkLoader.java:
```diff
@@ -0,0 +1,85 @@
+
+/**
+ * Loads an area around the player and renderViewEntity to the cache.
+ * The
```
Don't think you meant to leave that here.

ZacSharp commented on src/main/java/baritone/cache/ChunkLoader.java:
```diff
@@ -0,0 +1,85 @@
+        // Don't do anything if the world doesn't exist. Like in the main menu.
+        if (worldData == null) return;
+        loadAroundChunk(player.chunkCoordX, player.chunkCoordZ);
+        // Only load around the renderViewEntity if it's in a different chunk.
```
Maybe also don't do anything if the position didn't change since last tick?
Don't know whether the overhead of reloading every tick is more than the overhead of checking this.

ZacSharp commented on src/main/java/baritone/Baritone.java:
```diff
@@ -120,6 +121,9 @@
         this.selectionManager = new SelectionManager(this);
         this.commandManager = new CommandManager(this);
+
+        ChunkLoader chunkLoader = new ChunkLoader(this, 2); // load 3 by 3 area of chunks around the player.
```
#324 suggested loading 3 by 3 *regions* around the player, not chunks.
Loading 3 by 3 chunks changes almost nothing, it now just loads the next region 16 blocks before you reach the region border instead of when you reach it.

ZacSharp commented on src/main/java/baritone/cache/ChunkLoader.java:
```diff
@@ -0,0 +1,85 @@
+        if (worldData == null) return;
+        loadAroundChunk(player.chunkCoordX, player.chunkCoordZ);
+        // Only load around the renderViewEntity if it's in a different chunk.
+        if (player.chunkCoordX != renderViewEntity.chunkCoordX || player.chunkCoordZ != renderViewEntity.chunkCoordZ) {
```
You should compare region coords here because you can only load entire regions at once.

UPSman123 commented:
Thanks for the code review.
I was a bit confused about the difference between regions and chunks but I properly looked into it and I get it now. I changed it to work with regions instead.
I also added a check to see if the player moved to a different region. This won't prevent regions around the renderViewEntity from being loaded if it is in a different region but in that case it's only a few hashmap lookups and shouldn't influence performance significantly. I could add code to check for the renderViewEntity as well but I feel like it will just make things more messy without any real gain.

ZacSharp reviewed (changes requested):
Just two oversights and some formatting

UPSman123 commented:
I fixed it. Thanks again for looking it over.

ZacSharp reviewed (approved):
Looks good to me.
